### PR TITLE
OBPIH-6232 show hyperlink to source edit and stockcard on sources list page when user has sufficient roles

### PR DESCRIPTION
**Pull request by drodzewicz**
An additional case that I had to include in this ticket was conditionally rendered link to product sources edit if the user has sufficient roles (user is at least an admin and has product manager roles).

I have decided to create a custom hook `useUserHasPermission` to encapsulate the logic of retrieving user role data from redux and performing appropriate checks.

I also had to return location roles in the AppContext since I noticed that in a previous version of `hasPermission` util function we don't check for location roles on front end, only global roles.

### CODE REVIEW

File: src/js/hooks/useUserHasPermissions.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+   * @type {function(RoleType[]): boolean}
+   */
+  const hasEveryRole = useCallback((roles) => {
+    const userRoles = Array.isArray(currentUser?.roles) ? currentUser?.roles : [];
+    const userLocationRoles = Array.isArray(currentLocationRoles) ? currentLocationRoles : [];
+
+    const allUserRoles = new Set([...userRoles, ...userLocationRoles]);
+
+    return _.every(roles, (role) => allUserRoles.has(role));
+  }, [currentUser]);
```
Comment (kchelstowski) [inline src/js/hooks/useUserHasPermissions.js:43]: shouldn't the `currentLocation` also be a dependency?

Comment (drodzewicz) [reply src/js/hooks/useUserHasPermissions.js:43]: yes, true.
And currentUser dependancy should be `[currentUser.id]`

Comment (kchelstowski) [reply src/js/hooks/useUserHasPermissions.js:43]: exactly :D and the `currentLocation` -> `currentLocation?.id`

**Review by kchelstowski — APPROVED** [review]

**Review by jmiranda — COMMENTED** [review]
I'll let @awalkowiak decide on what to do here but I recommend moving the auth logic to the server and making it configurable.

File: src/js/hooks/useUserHasPermissions.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import RoleType from 'consts/roleType';
+
+/**
+ * Returns true or false if current logged-in user has all the roles specified in the parameters.
+ * The function checks if user has either global or location based roles.
+ * @param minRequiredRole {RoleType} - minimum required role
+ * @param supplementalRoles {RoleType[]} [d=[]] - list of supplemental required roles
+ * @returns {boolean}
+ */
+const useUserHasPermissions = ({ minRequiredRole, supplementalRoles = [] }) => {
```
Comment (awalkowiak) [inline src/js/hooks/useUserHasPermissions.js:15]: Should this be a replacement for the `permisionUtil` that we already have? Is there a case that we want to have both? I think we should have only one thing if that's possible.

File: src/js/components/productSupplier/ProductSupplierListTable.jsx [hunk 3]
```diff
@@ -30,6 +32,11 @@ const ProductSupplierListTable = ({ filterParams }) => {
     exportProductSuppliers,
   } = useProductSupplierActions({ fireFetchData });
 
+  const isUserAdminWithProductManager = useUserHasPermissions({
```
Comment (awalkowiak) [inline src/js/components/productSupplier/ProductSupplierListTable.jsx:35]: Let's change it to the suggested convention `canManageProducts` and let's stick to that version of variable names moving forward.

**Review by awalkowiak — CHANGES_REQUESTED** [review]
@drodzewicz see my comments for the change requests (at least change the name of variable, and do a second one if will be a quick one, otherwise let me know and I'll put it under the TD ticket that I am going to create for this one)

@jmiranda I agree with you here, that it would be better to handle this on the backend, but on the other hand if we'd need to do a separate property for each combination of `canDoSomething` this feels odd to me too. I prefer moving this froward as is and do a `Spike` together with future permissions/security refactor to investigate how we should and want to resolve this. I am going to create a TD ticket and we can discuss it further on the tech huddle.

**Review by awalkowiak — APPROVED** [review]

File: grails-app/controllers/org/pih/warehouse/api/ApiController.groovy [hunk 4]
```diff
@@ -144,6 +144,7 @@ class ApiController {
         User user = User.get(session?.user?.id)
         Location location = Location.get(session.warehouse?.id)
         String highestRole = user.getHighestRole(location)
+        List currentLocationRoles = user.getRolesByCurrentLocation(location)?.roleType*.name()
```
[the remark this answers is not in the document]
Comment (awalkowiak) [reply grails-app/controllers/org/pih/warehouse/api/ApiController.groovy:147]: I created this ticket for this one https://pihemr.atlassian.net/browse/OBPIH-6254

**Pull request merged**